File: 029InvertBinaryTree/invertBinaryTree.py
# Solution 1 - iteration
# O(n) time / O(n) space

def invertBinaryTree(tree):
    queue = [tree]
    while len(queue):
        currentNode = queue.pop(0)
        if currentNode is None:
            continue
        swapLeftAndRight(currentNode)
        queue.append(currentNode.left)
        queue.append(currentNode.right)


def swapLeftAndRight(tree):
    tree.left, tree.right = tree.right, tree.left

# Solution 2 - recursion
# O(n) time / O(d) space
# d - depth of the binary tree


def invertBinaryTree(tree):
    if tree is None:
        return
    swapLeftAndRight(tree)
    invertBinaryTree(tree.left)
    invertBinaryTree(tree.right)


def swapLeftAndRight(tree):
    tree.left, tree.right = tree.right, tree.left


# This is the class of the input binary tree.

class BinaryTree:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

File: 029InvertBinaryTree/invertBinaryTree.js
// Solution
// O(n) time / O(n) space

const invertBinaryTree = (tree) => {
  const queue = [tree];
  while (queue.length) {
    const currentNode = queue.shift();
    if (currentNode === null) continue;
    swapLeftAndRight(currentNode);
    queue.push(currentNode.left);
    queue.push(currentNode.right);
  }
};

const swapLeftAndRight = (tree) => {
  const left = tree.left;
  tree.left = tree.right;
  tree.right = left;
};

// Solution 2 - recursion
// O(n) time / O(d) space
// d - depth of binary tree

const invertBinaryTree = (tree) => {
  if (tree === null) return;
  swapLeftAndRight(tree);
  invertBinaryTree(tree.left);
  invertBinaryTree(tree.right);
};

const swapLeftAndRight = (tree) => {
  const left = tree.left;
  tree.left = tree.right;
  tree.right = left;
};

// This is the class of the input binary tree.

class BinaryTree {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}
